### ISSUE TITLE
Granular sanitization tools

### ISSUE THREAD

**Issue by dts** [PostHog/posthog-js#75]
We're building with posthog-js (note: huge fans overall) and rapidly found out how much our user-specified text is winding up in autocaptured events.  This makes things difficult to follow in the paths diagram.  We tried the `ph-no-capture` and realized that this prevents any events from being captured.  Really, we just don't want user-specific text anywhere in the analytics system.  So far, you're already stripping some stuff out automatically (e.g. credit card numbers), but we want to go further.  To do so, we added a `sanitize_properties` callback (https://github.com/PostHog/posthog-js/pull/74), which lets us do everything we need.  In addition to this callback, some way of substituting a generic label for user-specified text would give us most of what we need (and reduce the complexity of our callback by ~80%).

**Comment by mariusandra** [comment]
Hey @dts ! Cool feature and thanks for the PR!

Before merging it, could you give a few examples of the type of data you see captured? Are these e.g. `input` text element values (bug)... or you're storing the user data on e.g. a `<button>${username}</button>` that gets captured (feature)? What does your own `sanitize_properties` implementation look like, so we could provide better docs?

**Comment by dts** [comment]
All of the built-in sanitization is working nicely (i.e. inputs, CC numbers), but `<button>{username}</button>` gets captured (or <h5>{money}</h5> that's part of a navigating/clickable element).  Our sanitize_properties looks like this:

```
const sanitizeIDs = string =>
  string.replace(
    /([^a-z0-9]|^)([a-z0-9]{20})([^a-z0-9]|$)/g,
    (m, a, b, c) => `${a}${'X'.repeat(20)}${c}`
  )

const sanitizeNumbers = string => string.replace(/[\d,]+/g, '<num>')

const SANITIZE_ALLOWLIST = [
  '$user_id',
  'distinct_id',
  'token',
  '$screen_height',
  '$screen_width',
  '$lib_version',
  '$insert_id',
  '$device_id',
  '$ce_version',
  '$browser_version',
  '$time',
  'classes',
  'attr__class',
  'nth_child',
  'nth_of_type',
  'tag_name',
]

export const sanitize = object => {
  if (Array.isArray(object)) {
    return object.map(sanitize)
  }
  if (typeof object == 'object') {
    let r = {}
    for (let key in object) {
      if (SANITIZE_ALLOWLIST.includes(key)) r[key] = object[key]
      else r[key] = sanitize(object[key], key)
    }
    if (r.$elements) {
      for (let e of r.$elements) {
        if (e['attr__data-sensitive-content']) {
          e.$el_text = `[${e['attr__data-sensitive-content']}]`
        }
      }
    }
    return r
  }
  if (typeof object == 'string') {
    return sanitizeNumbers(sanitizeIDs(object))
  }
  if (typeof object == 'number') {
    return '<number>'
  }
}
```

**Comment by mariusandra** [comment]
Thanks for this!
I merged PR #74 and will release it as `1.4.0` once we review PR #71 as well. Should be later today.

**Comment by dts** [comment]
Thanks!
